tests/Unit/Domain/CoordinateMaps/TimeDependent/Test_Shape.cpp

```diff
@@ -0,0 +1,443 @@
+    std::array<DataVector, 3> target_points, std::array<double, 3> center,
+    size_t l_max, size_t m_max,
+    const std::vector<std::vector<std::complex<double>>>& coefs,
+    TransitionFunc transition_func) {
+  const size_t num_points = target_points[0].size();
+
+  const auto centered_coords = target_points - center;
+
+  const DataVector target_thetas =
+      atan2(hypot(centered_coords[0], centered_coords[1]), centered_coords[2]);
+  const DataVector target_phis = atan2(centered_coords[1], centered_coords[0]);
+
+  DataVector angular_part(num_points);
+
+  for (size_t i = 0; i < num_points; ++i) {
+    angular_part[i] = evaluate_harmonic_expansion(
+        identity, gsl::at(target_thetas, i), gsl::at(target_phis, i), coefs,
+        l_max, m_max);
+  }
+  const auto spatial_part = transition_func(centered_coords);
+  return target_points - centered_coords * angular_part * spatial_part;
+}
+
+template <typename TransitionFunc>
+void test_analytical_solution(TransitionFunc transition_func, size_t l_max,
+                              size_t m_max, size_t num_points,
+                              gsl::not_null<std::mt19937*> generator) {
+  std::array<double, 3> center = generate_random_point(generator);
+  std::array<DataVector, 3> target_data =
+      generate_random_targets(num_points, generator);
+  const auto random_coefs = generate_random_coefs(l_max, m_max, generator);
+
+  FunctionsOfTimeMap functions_of_time{};
+  double time{};
+  auto map = CoordinateMaps::TimeDependent::Shape<TransitionFunc>{};
+  const auto spherepack_coefs =
+      convert_coefs_to_spherepack(random_coefs, l_max, m_max);
+  generate_random_map_time_and_f_of_time<TransitionFunc>(
+      make_not_null(&map), make_not_null(&time),
+      make_not_null(&functions_of_time), l_max, m_max, center, transition_func,
+      spherepack_coefs, generator);
+  const auto mapped_result = map(target_data, time, functions_of_time);
+  const auto analytical_result = calculate_analytical_map(
+      target_data, center, l_max, m_max, random_coefs, transition_func);
+  CHECK_ITERABLE_APPROX(mapped_result, analytical_result);
+}
+
+// calculate the Jacobian using spherical harmonics directly
+template <typename TransitionFunc>
+tnsr::Ij<DataVector, 3, Frame::NoFrame> calculate_analytical_jacobian(
+    const std::array<DataVector, 3>& target_points,
+    const std::array<double, 3>& center, size_t l_max, size_t m_max,
+    const std::vector<std::vector<std::complex<double>>>& coefs,
+    TransitionFunc transition_func) {
+  const size_t num_points = target_points[0].size();
+  const auto centered_coords = target_points - center;
+  const DataVector target_thetas =
+      atan2(hypot(centered_coords[0], centered_coords[1]), centered_coords[2]);
+  const DataVector target_phis = atan2(centered_coords[1], centered_coords[0]);
+  DataVector angular_part(num_points);
+  DataVector theta_gradient(num_points);
+  DataVector phi_gradient(num_points);
```

Comment:
Instead of phi_gradient would it be possible to directly evaluate 1/sin(theta) d/dphi ?    Then all the dividing by sin(theta) would go away, and you wouldn't need to make sure your random numbers are away from the pole.